File: packages/contracts/prompt/resolver.ts
/**
 * Loop 2 Phase 2 — IWO3 prompt-resolution service (TypeScript).
 *
 * Pure function (no DB, no network, no clock). Given a pre-fetched set of
 * prompt layers and an optional operator override, produce the rendered
 * text, a SHA-256 `render_hash`, and a provenance echo of the layer ids.
 *
 * The TypeScript side is the reference implementation. The Python mirror at
 * `apps/api-fastapi/prompt/resolver.py` must produce byte-identical output
 * for the same input; the parity contract test
 * (`tests/contract/prompt-resolver-parity.test.ts`) enforces this across
 * eight fixture inputs.
 *
 * Policy (per IWO3_LOOP_2_APPROVAL_DECISIONS §Q2, ADR-002):
 *   - override kind `safety` is HARD-REJECTED in v0. Calling code must
 *     not attempt safety/tenant/tool policy overrides.
 *   - `external_send` is allowed through the resolver but flagged on the
 *     output; the approval gate lives in the adapter layer (Loop 3).
 *   - `style` and `profile_swap` pass through with the override text
 *     appended as the final layer and `overrideRef` populated.
 *
 * Rendering algorithm (must be identical in Python):
 *   1. Collect present layers in order: base, product, profile,
 *      workflow, wo, override.text (if any).
 *   2. Normalize each layer:
 *        a. strip trailing spaces/tabs from each line (regex: `[ \t]+$`
 *           per line, multi-line mode).
 *        b. strip leading + trailing whitespace from the result.
 *   3. Join normalized layers with "\n\n".
 *   4. Strip leading + trailing whitespace from the final joined string.
 *   5. SHA-256 the UTF-8 bytes. Hex digest = `renderHash`.
 */

import { createHash } from "node:crypto";

export type OverrideKind = "style" | "profile_swap" | "safety" | "external_send";

export interface ResolverInput {
  layers: {
    base?: string;
    product?: string;
    profile: string;
    workflow?: string;
    wo?: string;
  };
  layerIds: {
    baseId?: string;
    productId?: string;
    profileId: string;
    workflowId?: string;
    woId?: string;
  };
  override?: {
    kind: OverrideKind;
    text: string;
    userId: string;
    reason: string;
  };
}

export interface ResolverOutput {
  renderedText: string;
  renderHash: string;
  layerIds: ResolverInput["layerIds"] & { overrideId?: string };
  overrideRef?: {
    kind: OverrideKind;
    userId: string;
    reason: string;
  };
  requiresApproval: boolean;
}

export class SafetyOverrideRejected extends Error {
  constructor() {
    super(
      "Override rejected: safety/tenant/tool overrides are disallowed in v0 " +
        "(IWO3_LOOP_2_APPROVAL_DECISIONS §Q2; ADR-002)."
    );
    this.name = "SafetyOverrideRejected";
  }
}

function normalize(text: string): string {
  // Strip trailing spaces/tabs from each line, then trim the full string.
  return text.replace(/[ \t]+$/gm, "").trim();
}

export function resolvePrompt(input: ResolverInput): ResolverOutput {
  if (input.override && input.override.kind === "safety") {
    throw new SafetyOverrideRejected();
  }

  const pieces: string[] = [];
  if (input.layers.base !== undefined) pieces.push(normalize(input.layers.base));
  if (input.layers.product !== undefined) pieces.push(normalize(input.layers.product));
  pieces.push(normalize(input.layers.profile));
  if (input.layers.workflow !== undefined) pieces.push(normalize(input.layers.workflow));
  if (input.layers.wo !== undefined) pieces.push(normalize(input.layers.wo));
  if (input.override) pieces.push(normalize(input.override.text));

  // Drop empty pieces so an empty layer doesn't inject a blank line block.
  const nonEmpty = pieces.filter((p) => p.length > 0);

  const renderedText = nonEmpty.join("\n\n").trim();
  const renderHash = createHash("sha256")
    .update(renderedText, "utf-8")
    .digest("hex");

  const requiresApproval = input.override?.kind === "external_send";

  return {
    renderedText,
    renderHash,
    layerIds: input.layerIds,
    overrideRef: input.override
      ? {
          kind: input.override.kind,
          userId: input.override.userId,
          reason: input.override.reason,
        }
      : undefined,
    requiresApproval,
  };
}

File: apps/api-fastapi/prompt/resolver.py
"""Loop 2 Phase 2 — IWO3 prompt-resolution service (Python parity mirror).

Byte-identical output to `packages/contracts/prompt/resolver.ts`. The TS
side is the reference; parity is enforced by
`tests/contract/prompt-resolver-parity.test.ts` across eight fixture
inputs.

Algorithm (must match TS):
  1. Collect present layers in order: base, product, profile, workflow,
     wo, override.text (if any).
  2. Normalize each layer:
       a. strip trailing spaces/tabs from each line (regex [ \\t]+$
          multi-line).
       b. strip leading + trailing whitespace from the result.
  3. Drop empty normalized pieces.
  4. Join with "\\n\\n".
  5. Strip leading + trailing whitespace from the final joined string.
  6. SHA-256 the UTF-8 bytes; hex digest = renderHash.

Policy (IWO3_LOOP_2_APPROVAL_DECISIONS §Q2, ADR-002):
  - override kind "safety" is HARD-REJECTED.
  - "external_send" passes through with requires_approval=True.
  - "style" and "profile_swap" pass through normally.
"""

from __future__ import annotations

import hashlib
import re
from typing import Any


class SafetyOverrideRejected(ValueError):
    def __init__(self) -> None:
        super().__init__(
            "Override rejected: safety/tenant/tool overrides are "
            "disallowed in v0 (IWO3_LOOP_2_APPROVAL_DECISIONS \u00a7Q2; "
            "ADR-002)."
        )


_TRAILING_WS_PER_LINE = re.compile(r"[ \t]+$", flags=re.MULTILINE)


def _normalize(text: str) -> str:
    return _TRAILING_WS_PER_LINE.sub("", text).strip()


def resolve_prompt(inp: dict[str, Any]) -> dict[str, Any]:
    override = inp.get("override")
    if override and override.get("kind") == "safety":
        raise SafetyOverrideRejected()

    layers = inp.get("layers", {})
    layer_ids = inp.get("layerIds", {})

    pieces: list[str] = []
    if "base" in layers:
        pieces.append(_normalize(layers["base"]))
    if "product" in layers:
        pieces.append(_normalize(layers["product"]))
    pieces.append(_normalize(layers["profile"]))
    if "workflow" in layers:
        pieces.append(_normalize(layers["workflow"]))
    if "wo" in layers:
        pieces.append(_normalize(layers["wo"]))
    if override:
        pieces.append(_normalize(override["text"]))

    non_empty = [p for p in pieces if p]
    rendered = "\n\n".join(non_empty).strip()
    render_hash = hashlib.sha256(rendered.encode("utf-8")).hexdigest()

    requires_approval = bool(override and override.get("kind") == "external_send")

    out: dict[str, Any] = {
        "renderedText": rendered,
        "renderHash": render_hash,
        "layerIds": layer_ids,
        "requiresApproval": requires_approval,
    }
    if override:
        out["overrideRef"] = {
            "kind": override["kind"],
            "userId": override["userId"],
            "reason": override["reason"],
        }
    return out
